### ISSUE TITLE
Python API: dictionary instead of list for get_derivative

### ISSUE THREAD

**Issue by neXyon** [darioizzo/audi#17]
Instead of implementing #16:

To make the API consistent with the evaluate function, the get_derivative function could take a dictionary as parameter instead of a list.

For example, instead of

```
prod = np.sum([gdual(0, _, 1) for _ in 'xyz'])

index = list(np.zeros(prod.symbol_set_size))
index[2] = 1

print(prod.get_derivative(index))
```

you could do

```
prod = np.sum([gdual(0, _, 1) for _ in 'xyz'])

print(prod.get_derivative({'dz': 1}))
```

**Comment by bluescarni** [comment]
@neXyon This should be easy to do, we just need to decide if we want #16 or this (or both?).

**Comment by darioizzo** [comment]
I think we definitely want this as the audi application at low orders (say 1,2,3,4,5) and many variables would otherwise always have a get_derivative({0,0,0,0,0,0,0,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,0,0,2,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0})
which seems like not a good API.

We can wither implement it in piranha (we as in you :) or i Audi. Not sure what makes more sense .... probably piranha in case its a feature useful there too

**Comment by bluescarni** [comment]
@darioizzo Right. I was thinking this is also basically mandatory if we want to have sparse monomials with (say) thousands of variables.

**Comment by darioizzo** [comment]
Should I proceed to implement it in AuDi or we delay it up to when its changed upstream (sparse monomial + dictionary query of monomials)?

**Comment by bluescarni** [comment]
Go ahead. I don't know when I will have time to go for the sparse monomial stuff, it looks like you need this now.

**Comment by darioizzo** [comment]
Ok, I will implement what is basically a work around, I will accept a dictionary and build a vector ....
We can then remove it when the upstream method is implemented correctly.

**Comment by darioizzo** [comment]
Done. I will keep this opened though as to remind us that the current solution is just a workaround.

**Comment by darioizzo** [comment]
It is also probably better to use exclusively the dictionary as it allows to return 0 if the symbol is not found. Something that is extremely useful when you do not really know whats in the symbol set and still want to take a derivative.
